### PR TITLE
[DNM] rbd-mirror HA: create teuthology thrasher for rbd-mirror

### PR DESCRIPTION
**Pull request by trociny**
Fixes: http://tracker.ceph.com/issues/18753
Signed-off-by: Mykola Golub <mgolub@mirantis.com>

### CODE REVIEW

**Comment by trociny** [comment]
@dillaman This is an experimental version. I just wan to know your opinion.

So the idea is to allow rbd_mirror.sh and rbd_mirror_stress.sh tests to be run in "admin socket free" mode, when (the leader) admin socket is not known or daemons are running remotely. In this case some tests are skipped, for others helper functions that use admin sockets are modified to query pool mirror status instead.

But there is one frequently used function that I can't replace -- flush. Without it the test may wait forever for replay complete due to commit position is not flushed if residual io is below IN_FLIGHT_IO_LOW_WATER_MARK. As a workaround I try to run flush periodically in the thrasher, but this looks ugly and I am not sure about the effect. I tried to run the tests on teuthology [1], they failed and at I don't know the cause yet, most probably it is something wrong with my tests.

Still, I am thinking if we could avoid to run flush in our tests? For me it does not look very nice that the commit position may never reach its final state without external flush. Could we e.g. add some timer for this?

[1] http://pulpito.ceph.com/trociny-2017-02-08_16:19:54-rbd-wip-18753---basic-smithi/

**Comment by dillaman** [comment]
@trociny I was thinking of something more like [1][2][3] where we would have a dedicated test task for randomly thrashing rbd-mirror daemons. We want to not only thrash the leader / master daemon but also the follower / slave daemons. In a perfect world we would also always just run under the thrasher scenario instead of two separate scenarios -- perhaps they could be consolidated w/ a one deaemon per cluster and many daemons per cluster yaml. As for handling the asok operations, perhaps the admin socket helper command should be modified to cycle through all possible daemon asok files and retry the op X times w/ a delay between each loop to account for leader elections.  Alternatively, perhaps the "rbd mirror image status" command could return some uniquely identifying information to discover which rbd-mirror daemon is responsible for which image (ip address / client id). I think that would be a valuable addition regardless of whether or not it would be helpful for interacting w/ asok files.

[1] https://github.com/ceph/ceph/blob/master/qa/tasks/mon_thrash.py
[2] https://github.com/ceph/ceph/blob/master/qa/tasks/mds_thrash.py
[3] https://github.com/ceph/ceph/blob/master/qa/tasks/thrashosds.py

**Comment by trociny** [comment]
@dillaman

> I was thinking of something more like [1][2][3] where we would have a dedicated test task for randomly thrashing rbd-mirror daemons. We want to not only thrash the leader / master daemon but also the follower / slave daemons. 

Agreed. I just wanted to have some simple version so I could see potential issues from the start. A more elaborated one will require some work but this looks straightforward.

> In a perfect world we would also always just run under the thrasher scenario instead of two separate scenarios -- perhaps they could be consolidated w/ a one deaemon per cluster and many daemons per cluster yaml. 

This is what I hoped to achieve, but failed, most probably due too lack of experience with teuthology. I plan to try to cleanup this again later. Now I would like to solve the admin socket issue first.

> As for handling the asok operations, perhaps the admin socket helper command should be modified to cycle through all possible daemon asok files and retry the op X times w/ a delay between each loop to account for leader elections. Alternatively, perhaps the "rbd mirror image status" command could return some uniquely identifying information to discover which rbd-mirror daemon is responsible for which image (ip address / client id). I think that would be a valuable addition regardless of whether or not it would be helpful for interacting w/ asok files.

This looks racy, and I am not sure we can properly handle all cases. If not taking tests for asok operations, we need admin socket for these cases:

1) test an image replayer state (started, replaying, stopped)
2) start/stop an image replayer
3) flush am image replayer

For (2) and (3) it looks like doable -- iterate through socket files to find the leader, execute the command; if the command failed due to the command does not exist, retry.

For (1) it is not so simple, because actually it may be valid that the command does not exist because the image replayer is not started. I expect some additional work here in any case because with the thrasher the points where we are waiting for particular image replayer state become racy (e.g. we expect it to be stopped due to (after) some particular event, but it is actually stopped due new leader election). This needs some additional thoughts and actually I expect we could avoid using admin socket here, e.g. using only 'rbd mirror image status' command. The current version actually tries to use this approach but it looks like it needs additional work.

So I think we can use your approach for (2) and (3); how to properly deal with (1) is not clear yet.

On the other hand, from a user perspective, the operations:

1) find the leader instance
2) release the leader (acquire for particular instance)
3) start/stop/flush an image replayer

could be useful, and it would be not very convenient for users if these operations are available only via asok, so they need to iterate over all rbd-mirror nodes to find the leader (or instance that is currently responsible for the image). It would also difficult when writing a third party application for managing replication. I expect we will have a user request to provide a more convenient way (e.g. using the leader watcher messages).  Then may be we could start from this task, and then it would simplify further work on the thrasher?

**Comment by dillaman** [comment]
@trociny Just to clarify: we don't want to iterate over and find the leader since you would need to throw this work out once A/A is merged. The tests need to find the rbd-mirror instance that is responsible for a given image (which may not be the leader in the future) and execute the asok operation. It will definitely take careful planning to avoid races (i.e. look up instance that owns the image (loop if it cannot be found), execute the command (loop if command fails due to race).

File: qa/suites/rbd/mirror-thrasher/base [hunk 1]
```diff
@@ -0,0 +1 @@
+../mirror/base
```
Comment (amitkumar50) [inline qa/suites/rbd/mirror-thrasher/base]: @trociny (NIT) Please consider to make title of PR<50 characters
https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#title-of-pull-requests-and-title-of-commits

**Comment by dillaman** [comment]
Closing in favor of PR #21541

**Pull request closed without merging**